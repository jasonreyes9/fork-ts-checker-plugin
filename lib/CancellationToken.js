var ts = require('typescript');
var fs = require('fs');
var os = require('os');
var path = require('path');
var crypto = require('crypto');

function CancellationToken (cancellationFileName, isCancelled) {
  this.isCancelled = !!isCancelled;
  this.cancellationFileName = cancellationFileName || crypto.randomBytes(64).toString('hex');
  this.lastCancellationCheckTime = 0;
}
module.exports = CancellationToken;

CancellationToken.createFromJSON = function (json) {
  return new CancellationToken(
    json.cancellationFileName,
    json.isCancelled
  );
};

CancellationToken.prototype.toJSON = function () {
  return {
    cancellationFileName: this.cancellationFileName,
    isCancelled: this.isCancelled
  };
};

CancellationToken.prototype.isCancellationRequested = function () {
  if (this.isCancelled) {
    return true;
  }

  var time = Date.now();
  var duration = Math.abs(time - this.lastCancellationCheckTime);

  if (duration > 10) {
    // check no more than once every 10ms
    this.lastCancellationCheckTime = time;
    this.isCancelled = fs.existsSync(getCancellationFilePath(this.cancellationFileName));
  }

  return this.isCancelled;
};

CancellationToken.prototype.throwIfCancellationRequested = function () {
  if (this.isCancellationRequested()) {
    throw new ts.OperationCanceledException();
  }
};

CancellationToken.prototype.requestCancellation = function () {
  fs.writeFileSync(getCancellationFilePath(this.cancellationFileName), '');
};

CancellationToken.prototype.cleanupCancellation = function () {
  if (this.isCancelled) {
    fs.unlink(getCancellationFilePath(this.cancellationFileName));
  }
};

function getCancellationFilePath (cancellationFileName) {
  return path.join(os.tmpdir(), cancellationFileName);
}
