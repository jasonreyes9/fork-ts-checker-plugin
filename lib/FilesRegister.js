
function FilesRegister(oldRegister) {
  this.register = {};

  if (oldRegister) {
    oldRegister.forEach(function (fileName, fileEntry) {
      if (!fileEntry.modified) {
        // don't copy - just set reference to save memory
        this.register[fileName] = fileEntry;
      }
    }.bind(this));
  }
}
module.exports = FilesRegister;

FilesRegister.prototype.forEach = function (loop) {
  Object.keys(this.register).forEach(function (key) {
    loop(key, this.register[key]);
  }.bind(this));
};

FilesRegister.prototype.addFile = function (fileName) {
  this.register[fileName] = {
    modified: false,
    mtime: undefined,
    source: undefined,
    linted: false,
    lints: []
  };
};

FilesRegister.prototype.removeFile = function (fileName) {
  if (this.hasFile(fileName)) {
    delete this.register[fileName];
  }
};

FilesRegister.prototype.hasFile = function (fileName) {
  return this.register.hasOwnProperty(fileName);
};

FilesRegister.prototype.getFile = function (fileName) {
  if (!this.hasFile(fileName)) {
    throw new Error('File "' + fileName + '" not found in register.');
  }

  return this.register[fileName];
};

FilesRegister.prototype.ensureFile = function (fileName) {
  if (!this.hasFile(fileName)) {
    this.addFile(fileName);
  }
};

FilesRegister.prototype.setSource = function (fileName, source) {
  this.ensureFile(fileName);

  this.register[fileName].source = source;
};

FilesRegister.prototype.hasSource = function (fileName) {
  return this.hasFile(fileName) && !!this.getFile(fileName).source;
};

FilesRegister.prototype.getSource = function (fileName) {
  if (!this.hasSource(fileName)) {
    throw new Error('Cannot get source of "' + fileName + '" file.');
  }

  return this.getFile(fileName).source;
};

FilesRegister.prototype.setMtime = function (fileName, mtime) {
  this.ensureFile(fileName);

  if (this.register[fileName].mtime !== mtime) {
    this.register[fileName].modified = true;
    this.register[fileName].linted = false;
    this.register[fileName].lints = [];
    this.register[fileName].mtime = mtime;
  }
};

FilesRegister.prototype.getLints = function () {
  var lints = [];

  this.forEach(function (fileName, fileEntry) {
    lints.push.apply(lints, fileEntry.lints);
  });

  return lints;
};

FilesRegister.prototype.consumeLint = function (lint) {
  var fileName = lint.getFileName();

  this.ensureFile(fileName);
  this.register[fileName].linted = true;
  this.register[fileName].lints.push(lint);
};

FilesRegister.prototype.consumeLints = function (lints) {
  lints.forEach(function (lint) {
    this.consumeLint(lint);
  }.bind(this));
};
