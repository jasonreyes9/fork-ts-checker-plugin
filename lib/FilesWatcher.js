var chokidar = require('chokidar');
var path = require('path');
var startsWith = require('lodash.startswith');

function FilesWatcher(watchPaths, watchExtensions) {
  this.watchPaths = watchPaths;
  this.watchExtensions = watchExtensions;
  this.watchers = [];
  this.changeListeners = [];
  this.unlinkListeners = [];
}
module.exports = FilesWatcher;

FilesWatcher.prototype.isFileSupported = function(filePath) {
  return -1 !== this.watchExtensions.indexOf(path.extname(filePath));
};

FilesWatcher.prototype.watch = function () {
  this.watchers = this.watchPaths.map(function (watchPath) {
    return chokidar.watch(
      watchPath,
      { persistent: true }
    )
    .on('change', function(filePath, stats) {
      if (this.isFileSupported(filePath)) {
        this.changeListeners.forEach(function (changeListener) {
          changeListener(filePath, stats);
        });
      }
    }.bind(this))
    .on('unlink', function(filePath) {
      if (this.isFileSupported(filePath)) {
        this.unlinkListeners.forEach(function (unlinkListener) {
          unlinkListener(filePath);
        });
      }
    }.bind(this));
  }.bind(this));
};

FilesWatcher.prototype.isWatchingFile = function (filePath) {
  return (
    this.isWatching() &&
    this.watchPaths.some(function (watchPath) {
      return startsWith(filePath, watchPath);
    })
  );
};

FilesWatcher.prototype.isWatching = function () {
  return this.watchers.length > 0;
};

FilesWatcher.prototype.onChange = function (changeListener) {
  this.changeListeners.push(changeListener);

  return function unsubscribe() {
    this.changeListeners.filter(function (listener) {
      return listener !== changeListener;
    })
  }.bind(this);
};

FilesWatcher.prototype.onUnlink = function (unlinkListener) {
  this.unlinkListeners.push(unlinkListener);

  return function unsubscribe() {
    this.unlinkListeners.filter(function (listener) {
      return listener !== unlinkListener;
    })
  }.bind(this);
};
