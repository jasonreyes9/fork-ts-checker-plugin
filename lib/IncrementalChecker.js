var ts = require('typescript');
var fs = require('fs');
var path = require('path');
var endsWith = require('lodash.endswith');
var FilesRegister = require('./FilesRegister');
var FilesWatcher = require('./FilesWatcher');

function IncrementalChecker (programConfigFile, linterConfigFile, watchPaths) {
  this.programConfigFile = programConfigFile;
  this.linterConfigFile = linterConfigFile;
  this.watchPaths = watchPaths;
}
module.exports = IncrementalChecker;

IncrementalChecker.loadProgramConfig = function (configFile) {
  return ts.parseJsonConfigFileContent(
    ts.readConfigFile(configFile, ts.sys.readFile).config,
    ts.sys,
    path.dirname(configFile)
  );
};

IncrementalChecker.loadLinterConfig = function (configFile) {
  var tslint = require('tslint');

  return tslint.Configuration.loadConfigurationFromPath(configFile);
};

IncrementalChecker.createProgram = function (programConfig, register, watcher, oldProgram) {
  var host = ts.createCompilerHost(programConfig.options);
  var originGetSourceFile = host.getSourceFile;

  host.getSourceFile = function (filePath, languageVersion, onError) {
    // first check if watcher is watching file - if not - check it's mtime
    if (!watcher.isWatchingFile(filePath)) {
      var stats = fs.statSync(filePath);

      register.setMtime(filePath, stats.mtime);
    }

    // get source file only if there is no source in files register
    if (!register.hasSource(filePath)) {
      register.setSource(
        filePath,
        originGetSourceFile(filePath, languageVersion, onError)
      );
    }

    return register.getSource(filePath);
  };

  return ts.createProgram(
    programConfig.fileNames,
    programConfig.options,
    host,
    oldProgram // re-use old program
  );
};

IncrementalChecker.createLinter = function (program) {
  var tslint = require('tslint');

  return new tslint.Linter({ fix: false }, program);
};

IncrementalChecker.prototype.nextIteration = function () {
  this.register = new FilesRegister(this.register);

  if (!this.watcher) {
    this.watcher = new FilesWatcher(this.watchPaths, ['.ts', '.tsx']);

    // connect watcher with register
    this.watcher.onChange(function (filePath, stats) { this.register.setMtime(filePath, stats.mtime); }.bind(this));
    this.watcher.onUnlink(function (filePath) { this.register.removeFile(filePath); }.bind(this));

    this.watcher.watch();
  }

  if (!this.programConfig) {
    this.programConfig = IncrementalChecker.loadProgramConfig(this.programConfigFile);
  }

  if (!this.linterConfig && this.linterConfigFile) {
    this.linterConfig = IncrementalChecker.loadLinterConfig(this.linterConfigFile);
  }

  this.program = IncrementalChecker.createProgram(this.programConfig, this.register, this.watcher, this.program);
  this.linter = IncrementalChecker.createLinter(this.program);
};

IncrementalChecker.prototype.getDiagnostics = function (cancellationToken) {
  var diagnostics = this.program.getSemanticDiagnostics(undefined, cancellationToken);

  return diagnostics.map(function (diagnostic) {
    var position = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start);

    // normalize diagnostics
    return {
      type: 'typescript',
      code: diagnostic.code,
      category: ts.DiagnosticCategory[diagnostic.category].toLowerCase(),
      message: ts.flattenDiagnosticMessageText(diagnostic.messageText, '\n'),
      file: diagnostic.file.fileName,
      line: position.line,
      character: position.character
    };
  });
};

IncrementalChecker.prototype.getLints = function (cancellationToken) {
  this.register.forEach(function (fileName, fileEntry) {
    cancellationToken.throwIfCancellationRequested();

    if (!endsWith(fileName, '.d.ts') && !fileEntry.linted) {
      this.linter.lint(fileName, undefined, this.linterConfig);
      fileEntry.linted = true;
    }
  }.bind(this));

  this.register.consumeLints(
    this.linter.getResult().failures
  );

  var lints = this.register.getLints();

  return lints.map(function (lint) {
    var position = lint.getStartPosition().getLineAndCharacter();

    // normalize lints
    return {
      type: 'tslint',
      code: lint.getRuleName(),
      category: lint.getRuleSeverity(),
      message: lint.getFailure(),
      file: lint.getFileName(),
      line: position.line,
      character: position.character
    };
  });
};
