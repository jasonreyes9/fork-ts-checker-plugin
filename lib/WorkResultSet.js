var ts = require('typescript');

function WorkResultSet() {
  this.result = {};
}
module.exports = WorkResultSet;

WorkResultSet.prototype.set = function (key, result) {
  this.result[key] = result;
};

WorkResultSet.prototype.has = function (key) {
  return undefined !== this.result[key];
};

WorkResultSet.prototype.clear = function () {
  this.result = {};
};

WorkResultSet.prototype.done = function (keys) {
  return keys.every(function (key) {
    return this.has(key)
  }.bind(this));
};

WorkResultSet.prototype.merge = function () {
  var merged = {
    diagnostics: [],
    lints: []
  };

  Object.keys(this.result).forEach(function (key) {
    merged.diagnostics.push.apply(merged.diagnostics, this.result[key].diagnostics);
    merged.lints.push.apply(merged.lints, this.result[key].lints);
  }.bind(this));

  merged.diagnostics = ts.sortAndDeduplicateDiagnostics(merged.diagnostics);

  return merged;
};
