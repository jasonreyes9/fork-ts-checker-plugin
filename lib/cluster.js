var cluster = require('cluster');
var process = require('process');
var path = require('path');

var WorkResult = require('./WorkResult');
var NormalizedMessage = require('./NormalizedMessage');

// setup master
cluster.setupMaster({
  exec: path.join(__dirname, 'service.js'),
  args: ['--max-old-space-size=' + process.env.MEMORY_LIMIT]
});

// fork workers...
var division = parseInt(process.env.WORK_DIVISION);

for (var i = 0; i < division; i++) {
  cluster.fork({ WORK_NUMBER: i });
}

var workerIds = Object.keys(cluster.workers);
var result = new WorkResult(workerIds);

process.on('message', function (message) {
  // broadcast message to all workers
  workerIds.forEach(function (workerId) {
    cluster.workers[workerId].send(message);
  });

  // clear previous result set
  result.clear();
});

// listen to all workers
workerIds.forEach(function (workerId) {
  cluster.workers[workerId].on('message', function (message) {
    // set result from worker
    result.set(
      workerId,
      {
        diagnostics: message.diagnostics.map(NormalizedMessage.createFromJSON),
        lints: message.lints.map(NormalizedMessage.createFromJSON)
      }
    );

    // if we have result from all workers, send merged
    if (result.hasAll()) {
      var merged = result.reduce(
        function (merged, result) {
          return {
            diagnostics: merged.diagnostics.concat(result.diagnostics),
            lints: merged.lints.concat(result.lints)
          };
        },
        { diagnostics: [], lints: [] }
      );

      merged.diagnostics = NormalizedMessage.deduplicate(merged.diagnostics);
      merged.lints = NormalizedMessage.deduplicate(merged.lints);

      process.send(merged);
    }
  });
});
