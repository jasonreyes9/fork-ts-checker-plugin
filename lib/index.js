var path = require('path');
var process = require('process');
var childProcess = require('child_process');
var chalk = require('chalk');
var fs = require('fs');
var isString = require('lodash.isstring');
var CancellationToken = require('./CancellationToken');

/**
 * ForkTsCheckerWebpackPlugin
 * Runs typescript type checker and linter (tslint) on separate process.
 * This speed-ups build a lot.
 *
 * Options description in README.md
 */
function ForkTsCheckerWebpackPlugin (options) {
  this.tsconfig = options.tsconfig || './tsconfig.json';
  this.tslint = options.tslint === false ? false : options.tslint || './tslint.json';
  this.watch = isString(options.watch) ? [options.watch] : options.watch || [];
  this.blockEmit = !!options.blockEmit;
  this.ignoreDiagnostics = options.ignoreDiagnostics || [];
  this.ignoreLints = options.ignoreLints || [];
  this.logger = options.logger || console;
  this.silent = !!options.silent;

  this.tsconfigPath = undefined;
  this.tslintPath = undefined;
  this.watchPaths = [];
  this.compiler = undefined;
  this.colors = new chalk.constructor({
    enabled: options.colors === undefined ? true : !!options.colors
  });
  this.started = undefined;
  this.elapsed = undefined;
  this.cancellationToken = undefined;
  this.checkDone = false;
  this.compilationDone = false;
  this.diagnostics = [];
  this.lints = [];

  this.emitCallback = this.createNoopEmitCallback();
  this.doneCallback = this.createDoneCallback();
}
module.exports = ForkTsCheckerWebpackPlugin;

ForkTsCheckerWebpackPlugin.prototype.apply = function (compiler) {
  this.compiler = compiler;

  this.tsconfigPath = this.computeContextPath(this.tsconfig);
  this.tslintPath = this.tslint ? this.computeContextPath(this.tslint) : null;
  this.watchPaths = this.watch.map(this.computeContextPath.bind(this));

  // validate config
  var tsconfigOk = fs.existsSync(this.tsconfigPath);
  var tslintOk = !this.tslintPath || fs.existsSync(this.tslintPath);

  if (tsconfigOk && tslintOk) {
    this.pluginCompile();

    if (this.blockEmit) {
      this.pluginAfterEmit();
    } else {
      this.pluginDone();
    }
  } else {
    if (!tsconfigOk) {
      throw new Error(
        'Cannot find "' + this.tsconfigPath + '" file. Please check webpack and ForkTsCheckerWebpackPlugin configuration. \n' +
        'Possible errors: \n' +
        '  - wrong `context` directory in webpack configuration' +
        ' (if `tsconfig` is not set or is a relative path in fork plugin configuration)\n' +
        '  - wrong `tsconfig` path in fork plugin configuration' +
        ' (should be a relative or absolute path)'
      );
    }
    if (!tslintOk) {
      throw new Error(
        'Cannot find "' + this.tslintPath + '" file. Please check webpack and ForkTsCheckerWebpackPlugin configuration. \n' +
        'Possible errors: \n' +
        '  - wrong `context` directory in webpack configuration' +
        ' (if `tslint` is not set or is a relative path in fork plugin configuration)\n' +
        '  - wrong `tslint` path in fork plugin configuration' +
        ' (should be a relative or absolute path)\n' +
        '  - `tslint` path is not set to false in fork plugin configuration' +
        ' (if you want to disable tslint support)'
      );
    }
  }
};

ForkTsCheckerWebpackPlugin.prototype.computeContextPath = function (filePath) {
  return path.isAbsolute(filePath)
    ? filePath : path.resolve(this.compiler.options.context, filePath);
};

ForkTsCheckerWebpackPlugin.prototype.pluginCompile = function () {
  this.compiler.plugin('compile', function () {
    if (this.cancellationToken) {
      // request cancellation if there is not finished job
      this.cancellationToken.requestCancellation();
    }
    this.checkDone = false;
    this.compilationDone = false;

    this.started = process.hrtime();

    // create new token for current job
    this.cancellationToken = new CancellationToken();
    if (!this.service || !this.service.connected) {
      this.spawnService();
    }
    this.service.send(this.cancellationToken);
  }.bind(this));
};

ForkTsCheckerWebpackPlugin.prototype.pluginAfterEmit = function () {
  this.compiler.plugin('after-emit', function (compilation, callback) {
    this.emitCallback = this.createEmitCallback(compilation, callback);

    if (this.checkDone) {
      this.emitCallback();
    }

    this.compilationDone = true;
  }.bind(this));
};

ForkTsCheckerWebpackPlugin.prototype.pluginDone = function () {
  this.compiler.plugin('done', function () {
    if (this.checkDone) {
      this.doneCallback();
    } else {
      if (this.compiler) {
        this.compiler.applyPlugins(
          'fork-ts-checker-waiting',
          this.tslint !== false
        );
      }
      if (!this.silent && this.logger) {
        this.logger.info(
          this.tslint
            ? 'Type checking in progress...'
            : 'Type checking and linting in progress...'
        );
      }
    }

    this.compilationDone = true;
  }.bind(this));
};

ForkTsCheckerWebpackPlugin.prototype.spawnService = function () {
  this.service = childProcess.fork(
    path.resolve(__dirname, './service.js'),
    [],
    {
      env: {
        TSCONFIG: this.tsconfigPath,
        TSLINT: this.tslintPath,
        WATCH: this.watchPaths.join('|')
      }
    }
  );

  this.compiler.applyPlugins('fork-ts-checker-service-start');

  if (!this.silent && this.logger) {
    this.logger.info(
      this.tslint
        ? [
          'Starting type checking and linting service...',
          this.colors.grey(this.tsconfigPath),
          this.colors.grey(this.tslintPath)
        ].join('\n')
        : [
          'Starting type checking service...',
          this.colors.grey(this.tsconfigPath)
        ].join('\n')
    );
    if (this.watchPaths.length) {
      this.logger.info(
        'Watching:' +
        (this.watchPaths.length > 1 ? '\n' : ' ') +
        this.watchPaths
          .map(function (path) { return this.colors.grey(path); }.bind(this))
          .join('\n')
      );
    }
  }

  this.service.on('message', this.handleServiceMessage.bind(this));
  this.service.on('exit', this.handleServiceExit.bind(this));
};

ForkTsCheckerWebpackPlugin.prototype.handleServiceMessage = function (message) {
  if (this.cancellationToken) {
    this.cancellationToken.cleanupCancellation();
    // job is done - nothing to cancel
    this.cancellationToken = undefined;
  }

  this.checkDone = true;
  this.elapsed = process.hrtime(this.started);
  this.diagnostics = message.diagnostics;
  this.lints = message.lints;

  if (this.ignoreDiagnostics.length) {
    this.diagnostics = this.diagnostics.filter(function (diagnostic) {
      return this.ignoreDiagnostics.indexOf(diagnostic.code) === -1;
    }.bind(this));
  }

  if (this.ignoreLints.length) {
    this.lints = this.lints.filter(function (lint) {
      return this.ignoreLints.indexOf(lint.code) === -1;
    }.bind(this));
  }

  if (this.compilationDone) {
    this.blockEmit ? this.emitCallback() : this.doneCallback();
  }
};

ForkTsCheckerWebpackPlugin.prototype.handleServiceExit = function (code, signal) {
  if (signal === 'SIGABRT') {
    // probably out of memory :/
    if (this.compiler) {
      this.compiler.applyPlugins('fork-ts-checker-service-out-of-memory');
    }
    if (!this.silent && this.logger) {
      this.logger.error(
        this.colors.red('Type checking and linting aborted - probably out of memory.')
      );
    }
  }
};

ForkTsCheckerWebpackPlugin.prototype.createEmitCallback = function (compilation, callback) {
  return function emitCallback () {
    this.diagnostics.concat(this.lints).forEach(function (error) {
      var line = error.line + 1;
      var character = error.character + 1;
      // webpack message format
      var message = {
        rawMessage: (
          error.category.toUpperCase() + (error.type === 'typescript' ? ' TS' : ' ') + error.code + ': ' +
          error.message
        ),
        message: '(' + line + ',' + character + '): ' + error.message,
        location: {
          line: line,
          character: character
        },
        file: error.file
      };

      if (error.category === 'error') {
        compilation.errors.push(message);
      } else {
        compilation.warnings.push(message);
      }
    });

    callback();
  };
};

ForkTsCheckerWebpackPlugin.prototype.createNoopEmitCallback = function () {
  return function noopEmitCallback () {};
};

ForkTsCheckerWebpackPlugin.prototype.createDoneCallback = function () {
  return function doneCallback () {
    var elapsed = Math.round(this.elapsed[0] * 1E9 + this.elapsed[1]);

    if (this.compiler) {
      this.compiler.applyPlugins(
        'fork-ts-checker-done',
        this.diagnostics,
        this.lints,
        elapsed
      );
    }
    if (!this.silent && this.logger) {
      if (this.diagnostics.length || this.lints.length) {
        this.diagnostics.concat(this.lints).forEach(function (error) {
          var logColor = error.category === 'error' ? this.colors.red : this.colors.yellow;
          var logMethod = error.category === 'error' ? this.logger.error : this.logger.warn;

          logMethod(
            logColor(error.category + ' at ' + error.file) +
            '(' + this.colors.cyan(error.line + 1) + ',' + this.colors.cyan(error.character + 1) + '): '
          );
          logMethod(
            this.colors.grey((error.type === 'typescript' ? 'TS' : '') + error.code + ': ') +
            error.message + '\n'
          );
        }.bind(this));
      }
      if (!this.diagnostics.length) {
        this.logger.info(this.colors.green('No type errors found'));
      }
      if (!this.lints.length) {
        this.logger.info(this.colors.green('No lint errors found'));
      }
      this.logger.info('Type checking and linting time: ' + this.colors.bold(Math.round(elapsed / 1E6)) + 'ms');
    }
  };
};
