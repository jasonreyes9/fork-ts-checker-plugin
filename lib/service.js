var ts = require('typescript');
var process = require('process');
var IncrementalChecker = require('./IncrementalChecker');
var CancellationToken = require('./CancellationToken');

var checker = new IncrementalChecker(
  process.env.TSCONFIG,
  process.env.TSLINT,
  process.env.WATCH === '' ? [] : process.env.WATCH.split('|')
);

var diagnostics = [];
var lints = [];

function run (cancellationToken) {
  checker.nextIteration();

  try {
    diagnostics = checker.getDiagnostics(cancellationToken);
    lints = checker.getLints(cancellationToken);
  } catch (error) {
    if (error instanceof ts.OperationCanceledException) {
      return;
    }

    throw error;
  }

  if (!cancellationToken.isCancellationRequested()) {
    process.send({
      diagnostics: diagnostics,
      lints: lints
    });
  }
}

process.on('message', function (message) {
  run(CancellationToken.createFromJSON(message));
});
