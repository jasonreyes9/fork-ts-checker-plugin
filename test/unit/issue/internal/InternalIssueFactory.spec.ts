import { createIssueFromInternalError } from '../../../../lib/issue';

describe('[UNIT] issue/internal/InternalIssueFactory', () => {
  const ERROR_WITH_MESSAGE_AND_STACK = {
    message: 'Stack overflow - out of memory',
    stack: [
      `Security context: 0x35903c44a49 <JS Object>`,
      `1: walkFunctionDeclaration [<PROJECT_DIR>/node_modules/webpack/lib/Parser.js:~443] [pc=0xa07a14ec8ee] (this=0x59f67991119 <a Parser with map 0x71f2d115d49>,statement=0x3507a80af661 <a Node with map 0x71f2d1157c9>)`,
      `2: walkStatement [<PROJECT_DIR>/node_modules/webpack/lib/Parser.js:~348] [pc=0xa07a06dfc10] (this=0x59f6799111...`,
      ``,
      `FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - process out of memory`,
      `Abort trap: 6`
    ].join('\n')
  };
  const ERROR_WITHOUT_MESSAGE_AND_STACK = {
    toString() {
      return 'Error: Stack overflow - out of memory';
    }
  };

  it.each([[ERROR_WITH_MESSAGE_AND_STACK], [ERROR_WITHOUT_MESSAGE_AND_STACK]])(
    'creates Issue from Internal Error: %p',
    error => {
      const issue = createIssueFromInternalError(error);

      expect(issue).toMatchSnapshot();
    }
  );
});
