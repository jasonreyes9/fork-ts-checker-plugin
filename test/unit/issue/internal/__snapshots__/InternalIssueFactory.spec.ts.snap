// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[UNIT] issue/internal/InternalIssueFactory creates Issue from Internal Error: {"message": "Stack overflow - out of memory", "stack": "Security context: 0x35903c44a49 <JS Object>
1: walkFunctionDeclaration [<PROJECT_DIR>/node_modules/webpack/lib/Parser.js:~443] [pc=0xa07a14ec8ee] (this=0x59f67991119 <a Parser with map 0x71f2d115d49>,statement=0x3507a80af661 <a Node with map 0x71f2d1157c9>)
2: walkStatement [<PROJECT_DIR>/node_modules/webpack/lib/Parser.js:~348] [pc=0xa07a06dfc10] (this=0x59f6799111...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - process out of memory
Abort trap: 6"} 1`] = `
Object {
  "code": "INTERNAL",
  "message": "Stack overflow - out of memory",
  "origin": "internal",
  "severity": "error",
  "stack": "Security context: 0x35903c44a49 <JS Object>
1: walkFunctionDeclaration [<PROJECT_DIR>/node_modules/webpack/lib/Parser.js:~443] [pc=0xa07a14ec8ee] (this=0x59f67991119 <a Parser with map 0x71f2d115d49>,statement=0x3507a80af661 <a Node with map 0x71f2d1157c9>)
2: walkStatement [<PROJECT_DIR>/node_modules/webpack/lib/Parser.js:~348] [pc=0xa07a06dfc10] (this=0x59f6799111...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - process out of memory
Abort trap: 6",
}
`;

exports[`[UNIT] issue/internal/InternalIssueFactory creates Issue from Internal Error: {"toString": [Function toString]} 1`] = `
Object {
  "code": "INTERNAL",
  "message": "Error: Stack overflow - out of memory",
  "origin": "internal",
  "severity": "error",
  "stack": undefined,
}
`;
